### ISSUE TITLE
[Bug] Unit test infers the wrong data type when the first record has some NULL values

### ISSUE THREAD

**Issue by leesiongchan** [dbt-labs/dbt-redshift#821]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

```yml
unit_tests:
  - name: test_user_events_combined
    model: int_mycfg__user_events_combined
    given:
      - input: ref('stg_mycfg__user_events')
        rows:
          - {
              created_at: "2024-01-01 10:01:00",
              user_id: 1,
              is_closed: null, # <--- if we change both of these null values to true/false, the error goes away.
              is_validated: null, # <--- if we change both of these null values to true/false, the error goes away.
              user_action: created,
            }
          - {
              created_at: "2024-01-01 10:02:00",
              user_id: 1,
              is_closed: null,
              is_validated: true,
              user_action: validated,
            }
      - input: ref('stg_mycfg__user_snapshot')
        rows: []
    expect:
      rows:
        - { user_id: 1, is_closed: true, is_validated: true }
        - { user_id: 2, is_closed: null, is_validated: true }
        - { user_id: 3, is_closed: true, is_validated: null }
        - { user_id: 4, is_closed: null, is_validated: null }
        - { user_id: 5, is_closed: false, is_validated: true }
        - { user_id: 6, is_closed: true, is_validated: true }
```

When you run the test you will see error like this.

<img width="1151" alt="image" src="https://github.com/dbt-labs/dbt-core/assets/324418/9d7dbabf-8499-406f-b840-2793d531be69">

And here is the logs:

<img width="542" alt="image" src="https://github.com/dbt-labs/dbt-core/assets/324418/9ca15240-5b0e-4fda-9bf6-7d505984d8aa">

_Input build_

<img width="591" alt="image" src="https://github.com/dbt-labs/dbt-core/assets/324418/9d84ea6a-82bb-48ed-9fc6-68ea496006db">

_Expected result build_

Interestingly, the incorrect casting happen in the `dbt_internal_unit_test_expected` cte but not the input cte.

---

Our current workaround is to use sql to cast the correct type, eg.

```yml
select
    '2024-01-01 10:01:00' as created_at,
    1 as user_id,
    null::boolean as is_closed,
    null::boolean as is_validated,
    'created' as user_action
union all
select '2024-01-01 10:02:00', 1, null, true, 'validated'
```

### Expected Behavior

The test runner should be able to guess the data type correctly.

### Steps To Reproduce

1. Provide unit test data with some null values in the first record.
2. Run the test
3. ERROR!

### Relevant log output

<img width="542" alt="image" src="https://github.com/dbt-labs/dbt-core/assets/324418/9ca15240-5b0e-4fda-9bf6-7d505984d8aa">

_Input build_

<img width="591" alt="image" src="https://github.com/dbt-labs/dbt-core/assets/324418/9d84ea6a-82bb-48ed-9fc6-68ea496006db">

_Expected result build_

Interestingly, the incorrect casting happen in the `dbt_internal_unit_test_expected` cte but not the input cte.

### Environment

```markdown
- OS: Ubuntu 24.04 WSL
- Python: 3.12.2
- dbt: 1.8.0-rc1
```

### Which database adapter are you using with dbt?

redshift

**Comment by jnapoleon** [comment]
Hey @dbeatty10 I've attached the logs

Let me know if you would rather me create a new issue?
[log.txt](https://github.com/dbt-labs/dbt-redshift/files/15322838/log.txt)

Thanks

John

**Comment by dbeatty10** [comment]
Thanks for attaching your logs @jnapoleon 👍 

Could you also provide a minimal code example like the "reprex" within https://github.com/dbt-labs/dbt-redshift/issues/821#issuecomment-2108918024? That would allow us to determine if you are having the same issue or a different issue.

**Comment by jnapoleon** [comment]
@dbeatty10 I managed to resolve the issue, I don't think there was a bug with dbt. 

There was some strange behaviour getting the data types using information schema from the created temporary table. It thought the date type was varchar for the temporary table but in the real final output model it was a numeric

The way I solved it was explicitly casting the fields as numeric in the final output model

**Comment by joellabes** [comment]
I am also seeing this. Because I'm doing cross-platform unit tests for the audit helper package, I have a diff of what snowflake generated vs redshift: 
<img width="1635" alt="image" src="https://github.com/dbt-labs/dbt-redshift/assets/7335046/2bcf6b4f-d629-497f-9f03-024edf3c66c7">

When nulls are present alongside numeric values, dbt is casting everything as a varchar(1), which makes unions between expected and actual impossible. 

Here's the full file that ran in [this CI run](https://app.circleci.com/pipelines/github/dbt-labs/dbt-audit-helper/413/workflows/97f91684-a874-422c-a943-59e0fee85e98/jobs/406/steps)
[reworked_compare_identical_tables_multiple_null_pk.txt](https://github.com/dbt-labs/dbt-redshift/files/15359952/reworked_compare_identical_tables_multiple_null_pk.txt)

In my case, the input rows lead with the null but the expected output has a non-null value first. 

```yml
unit_tests:
  - name: reworked_compare_identical_tables_multiple_null_pk
    model: unit_reworked_compare
    
    given:
      - input: ref('unit_test_model_a')
        rows:
          - { "id": , "col1": "abc", "col2": "def" }
          - { "id": , "col1": "hij", "col2": "klm" }
          - { "id": 3, "col1": "nop", "col2": "qrs" }
      - input: ref('unit_test_model_b')
        rows:
          - { "id": , "col1": "abc", "col2": "def" }
          - { "id": , "col1": "hij", "col2": "klm" }
          - { "id": 3, "col1": "nop", "col2": "qrs" }
        
    expect:
      rows:
        - {"dbt_audit_row_status": 'identical', 'id': 3, dbt_audit_num_rows_in_status: 1}
        - {"dbt_audit_row_status": 'nonunique_pk', 'id': , dbt_audit_num_rows_in_status: 2}
        - {"dbt_audit_row_status": 'nonunique_pk', 'id': , dbt_audit_num_rows_in_status: 2}

    overrides:
      vars:
        reworked_compare__columns: ['id', 'col1', 'col2']
        reworked_compare__event_time:
        reworked_compare__primary_key_columns: ['id']
```

**Comment by VersusFacit** [comment]
I made a little reprex 💟 

```sql
-- models/staging/a.sql
{{ config(materialized="table") }}

select 1 as id, 'a' as col1
union all
select 2, 'b'
union all
select 3, 'c'
```

```sql
-- models/staging/b.sql
{{config(materialized="table")}}

select * from {{ ref('a') }}
```

```yaml
unit_tests:
  - name: test_simple

    model: b
    given:
      - input: ref('a')
        rows:
          - { "id":  , "col1": "d"}
          - { "id": , "col1": "e"}
          - { "id": 6, "col1": "f"}

    expect:
      rows:
        - {id:  , "col1": "d"}
        - {id:  , "col1": "e"}
        - {id: 6, "col1": "f"}

```

**Comment by VersusFacit** [comment]
Can't be merged without a unit test, so blocked until then. The work itself should be fast < half a point and to be taken care of posthaste on Monday

**Comment by VersusFacit** [comment]
To make this clear, we cannot solve this issue in all cases. There is one case where no single row has a non null value and Redshift's type interpolation degrades to `VARCHAR(1)`. We talked in house. There is no easy way with the existing framework to take care of this situation, which we imagine is exceedingly rare. Instead we flag a parsing error now which asks the user to write just one row with all non Null values. That way under the hood, this row can be used for first insertion into Redshift which helps the database determine the proper types, including for future `null` fields.

Note, I did learn that this is a seemingly Redshift specific behavior. Other dbs like postgres and snowflake seem to have some more sophisticated type interpolation behaviors.
